packages/babel-plugin-transform-class-properties/src/index.js

```diff
@@ -9,21 +9,212 @@ export default function({ types: t }) {
+      propNames[isStatic ? "staticProps" : "instanceProps"].push(path);
+
+      if (computed) continue;
+
+      const isPrivate = path.isClassPrivateProperty();
+      const name = isPrivate
+        ? key.id.name
+        : t.isIdentifier(key) ? key.name : key.value;
+      const seen =
+        propNames[
+          isPrivate
+            ? "privateProps"
+            : isStatic ? "publicStaticProps" : "publicProps"
+        ];
+
+      if (seen[name]) {
+        throw path.buildCodeFrameError("duplicate class field");
+      }
+      seen[name] = true;
+    }
+  }
+
+  const staticErrorVisitor = {
+    ClassBody(path) {
+      path.skip();
+    },
+
+    ClassProperty(path) {
+      const { computed, key, static: isStatic } = path.node;
+      if (computed || !isStatic) return;
+
+      const name = t.isIdentifier(key) ? key.name : key.value;
+      if (name === "prototype") {
+        throw path.buildCodeFrameError("illegal static class field");
       }
     },
+
+    PrivateName(path) {
+      const { parentPath, node } = path;
+      if (parentPath.isMemberExpression({ property: node, computed: false })) {
+        if (!this.privateProps[node.id.name]) {
+          throw path.buildCodeFrameError("unknown private property");
+        }
+
+        return;
+      }
+
+      if (parentPath.isClassPrivateProperty({ key: node })) return;
+
+      throw path.buildCodeFrameError(
+        `illegal syntax. Did you mean \`this.#${node.id.name}\`?`,
+      );
+    },
+  };
+
+  const privateNameRemapper = {
+    PrivateName(path) {
+      const { node, parent, parentPath, scope } = path;
+      if (node.id.name !== this.name) {
+        return;
+      }
+      if (!parentPath.isMemberExpression()) return;
+
+      const grandParentPath = parentPath.parentPath;
+      const { object } = parent;
+
+      let replacePath = parentPath;
+      let replaceWith = t.callExpression(this.get, [object, this.privateMap]);
+
+      if (
+        grandParentPath.isAssignmentExpression() ||
+        grandParentPath.isUpdateExpression()
+      ) {
+        const { node } = grandParentPath;
+        let assign;
+        let memo;
+        let postfix;
+
+        if (grandParentPath.isAssignmentExpression({ operator: "=" })) {
+          assign = node.right;
+        } else {
+          const { right, operator } = node;
+          memo = scope.maybeGenerateMemoised(object);
+
+          if (memo) {
+            replaceWith.arguments[0] = memo;
+            memo = t.assignmentExpression("=", memo, object);
+          }
+
+          if (grandParentPath.isUpdateExpression({ prefix: false })) {
+            postfix = scope.generateUidIdentifierBasedOnNode(parent);
+            scope.push({ id: postfix });
+            replaceWith = t.assignmentExpression(
+              "=",
+              postfix,
+              t.unaryExpression("+", replaceWith),
+            );
+          }
+
+          assign = t.binaryExpression(
```

Comment:
Very cute that this path works for both UpdateExpressions and AssignmentExpressions (e.g., the way `right` is not undefined only if it's an AssignmentExpression), but the duality was a bit confusing to read. It might benefit from either a comment or splitting into two separate branches.